cosmos/scheduler/cmd/scheduler.go

```diff
@@ -0,0 +1,247 @@
+	}
+	if !exists {
+		if err = s.createIngestionSecret(); err != nil {
+			panic(err.Error())
+		}
+	}
+	msgCh := make(chan interface{}, 1)
+	go func() {
+		for {
+			s.handler(msgCh)
+		}
+	}()
+	for {
+		if err := s.waitForOverlayUpdate(); err != nil {
+			log.Println(err)
+		} else {
+			msgCh <- "overlay update"
+		}
+	}
+}
+
+func (s *Scheduler) waitForOverlayUpdate() error {
+	collection := s.MongodbClient.Database("metadata").Collection("PENSIEVE")
+	ctx := context.Background()
+	cur, err := collection.Watch(ctx, mongo.Pipeline{})
+	if err != nil {
+		return err
+	}
+	if cur.Next(ctx) {
+		return nil
+	}
+	return cur.Err()
+}
+
+func (s *Scheduler) getIngestionSecret() (bool, error) {
+	log.Println("looking for ingestion credentials secret")
+	secrets, err := s.KubeClientset.CoreV1().
+		Secrets(s.Namespace).
+		List(metav1.ListOptions{LabelSelector: "ingestionCredentials=true"})
+	if err != nil {
+		return false, nil
+	}
+	if len(secrets.Items) > 0 {
+		log.Println("ingestion credentials secret found")
+		return true, nil
+	}
+	log.Println("ingestion credentials secret not found")
+	return false, nil
+}
+
+func (s *Scheduler) createIngestionSecret() error {
+	log.Println("creating ingestion credentials secret")
+	accessKey, secretKey, err := s.Pensieve.GetServiceAccountCredentials("md-ingestion")
+	if err != nil {
+		return err
+	}
+	_, err = s.KubeClientset.CoreV1().Secrets(s.Namespace).Create(&v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "cosmos-metadata-ingestion",
+			Namespace: s.Namespace,
+			Labels: map[string]string{
+				"ingestionCredentials": "true",
+			},
+		},
+		Type: "Opaque",
+		Data: map[string][]byte{
+			"accessKey": []byte(accessKey),
+			"secretKey": []byte(secretKey),
+		},
+	})
+	if err != nil {
+		return err
+	}
+	log.Println("ingestion credentials secret created succesfully")
+	return nil
+}
+
+func (s *Scheduler) getCosmosLocationBson() (bson.A, error) {
+	locations, err := s.Pensieve.GetLocationsWithTypes([]string{"nfs"})
+	if err != nil {
+		return nil, err
+	}
+	ret := bson.A{}
+	for _, location := range locations {
+		ret = append(ret, bson.D{
+			{"fullDocument.value.locationConstraint", location.Name},
+		})
+	}
+	return ret, nil
+}
+
+type BucketTransaction struct {
+	OperationType string `bson:"operationType"`
+	FullDocument  struct {
+		Value struct {
+			Name               string `bson:"name"`
+			OwnerDisplayName   string `bson:"ownerDisplayName"`
+			Deleted            bool   `bson:"deleted"`
+			LocationConstraint string `bson:"locationConstraint"`
+		} `bson:"value"`
+	} `bson:"fullDocument"`
+}
+
+func (s *Scheduler) watchBucketEvents(ch chan BucketTransaction, quit chan bool) {
+	locationBson, err := s.getCosmosLocationBson()
+	if err != nil {
+		log.Println(err)
+		return
+	}
+	if len(locationBson) == 0 {
+		return
+	}
+	collection := s.MongodbClient.Database("metadata").Collection("__metastore")
+	ctx := context.Background()
+	cur, err := collection.Watch(ctx, mongo.Pipeline{
```

Comment:
it might be safer to set `batchSize` for the watch to only return one document at a time.